src/policy/vmspace.rs

```diff
@@ -112,118 +144,104 @@ impl<VM: VMBinding> crate::policy::gc_work::PolicyTraceObject<VM> for VMSpace<VM
-            .then(|| Self::create_space(args, None));
-        Self {
-            inner,
-            args: args_clone,
+
+        if !vm_space_start.is_zero() {
+            space.add_external_pages(vm_space_start, vm_space_size);
         }
-    }
 
-    pub fn lazy_initialize(&mut self, start: Address, size: usize) {
-        assert!(self.inner.is_none(), "VM space has been initialized");
-        self.inner = Some(Self::create_space(&mut self.args, Some((start, size))));
+        space
+    }
 
-        self.common().initialize_sft(self.as_sft());
+    pub fn set_vm_region(&mut self, start: Address, size: usize) {
+        self.add_external_pages(start, size);
     }
 
-    fn create_space(
-        args: &mut CreateSpecificPlanArgs<VM>,
-        location: Option<(Address, usize)>,
-    ) -> ImmortalSpace<VM> {
-        use crate::util::heap::layout::vm_layout_constants::BYTES_IN_CHUNK;
+    pub fn add_external_pages(&self, start: Address, size: usize) {
+        let start = start.align_down(BYTES_IN_PAGE);
+        let end = (start + size).align_up(BYTES_IN_PAGE);
+        let size = end - start;
 
-        // If the location of the VM space is not supplied, find them in the options.
-        let (vm_space_start, vm_space_bytes) = location.unwrap_or((
-            *args.global_args.options.vm_space_start,
-            *args.global_args.options.vm_space_size,
-        ));
-        // Verify the start and the size is valid
-        assert!(!vm_space_start.is_zero());
-        assert!(vm_space_bytes > 0);
+        assert!(!start.is_zero());
+        assert!(size > 0);
 
-        // We only map on chunk granularity. Align them.
-        let vm_space_start_aligned = vm_space_start.align_down(BYTES_IN_CHUNK);
-        let vm_space_end_aligned = (vm_space_start + vm_space_bytes).align_up(BYTES_IN_CHUNK);
-        let vm_space_bytes_aligned = vm_space_end_aligned - vm_space_start_aligned;
+        let chunk_start = start.align_down(BYTES_IN_CHUNK);
+        let chunk_end = end.align_up(BYTES_IN_CHUNK);
+        let chunk_size = chunk_end - chunk_start;
 
         // For simplicity, VMSpace has to be outside our available heap range.
         // TODO: Allow VMSpace in our available heap range.
         assert!(Address::range_intersection(
-            &(vm_space_start_aligned..vm_space_end_aligned),
+            &(chunk_start..chunk_end),
             &crate::util::heap::layout::available_range()
         )
         .is_empty());
 
         debug!(
             "Align VM space ({}, {}) to chunk ({}, {})",
-            vm_space_start,
-            vm_space_start + vm_space_bytes,
-            vm_space_start_aligned,
-            vm_space_end_aligned
-        );
-
-        let space_args = args.get_space_args(
-            "vm_space",
-            false,
-            VMRequest::fixed(vm_space_start_aligned, vm_space_bytes_aligned),
+            start, end, chunk_start, chunk_end
```

Comment:
Maybe we should use `page_start` and `page_end` here, but `start` and `end` are still useful information to show.